adr/keycloak-adr.md

```diff
@@ -0,0 +1,236 @@
+- **TLS/SSL**: All communication between Keycloak, the event listener, NATS, and the service listeners will be encrypted using TLS/SSL to ensure data confidentiality and integrity. This will be achieved by utilizing istio for mutual TLS between services. This will  Strict mTLS will be enforced between services to ensure that only authorized entities can communicate with each other.
+
+- **Restricted Access**: Access to the event listener, NATS, and the service listeners will be restricted to authorized entities only using Kubernetes Network Policies and istio authorization policies.
+
+#### Implications
+- **Security and Compliance**: Aims to heighten security and compliance standards, with an understanding that detailed RMF documentation and compliance will develop over time.
+- **User Experience**: Foresees significant improvements in user experience by offering streamlined access across services.
+- **License Management**: Emphasizes the critical role of GitLab Premium for accessing sophisticated authentication features.
+- **Future Enhancements**: Envisions upcoming developments, especially the integration of Keycloak with GitLab groups, to further refine access controls and permission management.
+
+### Audit trail
+The integration of Keycloak with GitLab will also address the need for an audit trail, ensuring that all user access and deprovisioning events are logged and auditable. This is a critical requirement for security and compliance.
+
+#### Considerations
+- **Audit Logging**: All user access and deprovisioning events should be logged and stored securely to ensure that they are auditable.
+- **Security and Compliance**: This capability is critical for security and compliance, ensuring that all user access and deprovisioning events are logged and auditable.
+- **Reliability and Scalability**: The audit logging process should be reliable and scalable, ensuring that all user access and deprovisioning events are logged and auditable as the organization grows.
+
+#### Design Overview
+The cornerstone of our design strategy is to fortify security and compliance measures through a robust and scalable audit logging framework. This framework is designed to meticulously record and securely preserve all user access and deprovisioning activities, thereby facilitating comprehensive auditability and detailed reporting. In its initial deployment, our focus will be on establishing three foundational components. Direct event capture via Keycloak's Service Provider Interface (SPI) and the utilization of the NATS system for creating an append-only log structure are earmarked for subsequent integration to enhance the audit log's reliability and authenticity. Furthermore, we plan to incorporate hash chaining within the append-only log to guarantee data immutability and offer a formidable barrier against data manipulation attempts. These advancements are anticipated to significantly augment the audit log's security posture in future iterations.
```

Comment:
This is a lot and could be shortened significantly while not losing meaning- it mentions "three foundational components" which aren't very clear (at least to me) - Maybe they are Keycloak's SPI, NATS, and hash-chaining?  Are we going to write custom code to do the hash chaining?  What drives that requirement?